transforms/images/roi-relabel-plugin/src/polus/plugins/transforms/images/roi_relabel/__main__.py

```diff
@@ -0,0 +1,104 @@
+
+import concurrent.futures
+import json
+import logging
+import pathlib
+import time
+import typing
+
+import filepattern as fp
+import typer
+
+from polus.plugins.transforms.images.roi_relabel import methods, utils
+
+logging.basicConfig(
+    format="%(asctime)s - %(levelname)s - %(name)s - %(funcName)s - %(message)s",
+    datefmt="%d-%b-%y %H:%M:%S",
+)
+logger = logging.getLogger("polus.plugins.transforms.images.roi_relabel")
+logger.setLevel(utils.constants.POLUS_LOG)
+
+app = typer.Typer()
+
+
+@app.command()
+def _main(
+    inp_dir: pathlib.Path = typer.Option(
+        ..., "--inpDir", help="Input directory with labeled images."
+    ),
+    out_dir: pathlib.Path = typer.Option(
+        ..., "--outDir", help="Directory where output images should be written."
+    ),
+    method: methods.Methods = typer.Option("contiguous", help="to use for relabeling"),
+    pattern: str = typer.Option(
+        ".*", "--filePattern", help="filepattern for images in inpDir."
+    ),
+    seed: typing.Optional[int] = typer.Option(
+        None, help="Random seed to use for reproducibility."
+    ),
+    preview: bool = typer.Option(
+        False, help="Output a JSON preview of files generated by this tool."
+    ),
+):
+    inp_dir = inp_dir.resolve()
+    assert inp_dir.exists(), f"Path not found {inp_dir}"
+    if inp_dir.joinpath("images").is_dir():
+        inp_dir = inp_dir.joinpath("images")
+
+    out_dir = out_dir.resolve()
+    assert out_dir.exists(), f"Path not found {out_dir}"
+
+    assert inp_dir != out_dir, "inpDir and outDir must be different."
+
+    logger.info(f"inpDir = {inp_dir}")
+    logger.info(f"outDir = {out_dir}")
+    logger.info(f"{method = }")
+    logger.info(f"filePattern = {pattern}")
+    logger.info(f"{seed = }")
+
+    if preview:
+        with open(out_dir.joinpath("preview.json"), "w") as writer:
+            json.dump({"filePattern": pattern}, writer, indent=2)
+        return
+
+    if seed is not None:
+        utils.helpers.seed_everything(seed)
+
+    start = time.perf_counter()
+    logger.info("Getting started ...")
+
+    image_paths = [
+        pathlib.Path(path[1][0]) for path in fp.FilePattern(str(inp_dir), pattern)()
+    ]
+    with concurrent.futures.ProcessPoolExecutor(
+        max_workers=utils.constants.NUM_THREADS
+    ) as executor:
+        futures: list[concurrent.futures.Future[bool]] = list()
+
+        for inp_path in image_paths:
+            out_path = out_dir.joinpath(utils.helpers.replace_extension(inp_path.name))
+            futures.append(
+                executor.submit(
+                    methods.relabel,
+                    inp_path,
+                    out_path,
+                    method,
+                )
+            )
+
+        done, not_done = concurrent.futures.wait(futures, 10)
+        while len(not_done) > 0:
+            logger.info(f"Progress {100 * len(done) / len(futures):6.2f}% ...")
+            done, not_done = concurrent.futures.wait(futures, 10)
```

Comment:
As an alternative to this, check out Hamdah's pluging that uses tqdm with a minimum reporting interval. I think it might be cleaner than this. This is just a recommendation and is not required for acceptance of the PR.

Reply:
Done